### PR TITLE
Remove redundant getTable call during getPartition/getPartitionByNames

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java [hunk 1]
```diff
@@ -311,6 +305,19 @@ private void alterTable(HiveIdentity identity, String databaseName, String table
         return resultBuilder.build();
     }
 
+    private Function<org.apache.hadoop.hive.metastore.api.Partition, Partition> getFromMetastoreApiPartition(Table table)
+    {
+        boolean isAvroTableWithSchemaSet = ThriftMetastoreUtil.isAvroTableWithSchemaSet(table);
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java]: this actually fixes a bug

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java [hunk 2]
```diff
@@ -113,7 +115,12 @@ private void doCreateEmptyPartition(ConnectorSession session, String schema, Str
                 .map(String.class::cast)
                 .collect(toImmutableList());
 
-        if (metastore.getPartition(new HiveIdentity(session), schema, table, partitionStringValues).isPresent()) {
+        Optional<Table> table = metastore.getTable(new HiveIdentity(session), schema, tableName);
+        if (!table.isPresent()) {
+            throw new PrestoException(NOT_FOUND, "Table not found");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java]: - include table name in the message
- use `.orElseThrow`:

```
metastore.getTable(new HiveIdentity(session), schema, tableName)
  .orElseThrow
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java [hunk 3]
```diff
@@ -84,6 +101,7 @@ public String toString()
     {
         return toStringHelper(this)
                 .add("hiveTableName", hiveTableName)
+                .add("table", table)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: `table.orElse(null)`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 4]
```diff
@@ -626,8 +625,7 @@ private void invalidateTableStatisticsCache(String databaseName, String tableNam
     {
         return delegate.getPartition(
                 partitionName.getIdentity(),
-                partitionName.getKey().getHiveTableName().getDatabaseName(),
-                partitionName.getKey().getHiveTableName().getTableName(),
+                partitionName.getKey().getTable().get(),
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java]: What if `! partitionName.getKey().getTable().isPresent()`?
What if it is stale?

**Comment by sopel39** [comment]
> I am concerned about potential inconsistency implications of this change. (I am not saying there are any.)

I don't think such design was intentional. In some cases caller does have `Table` at hand, but doesn't use it

**Comment by findepi** [comment]
> I don't think such design was intentional. In some cases caller does have `Table` at hand, but doesn't use it

I understand. But now the `Table` becomes bound in the cache key, extending object lifecycle.

And this leads to another concern -- memory footprint of the `Table` objects referenced by the cache.

**Comment by sopel39** [comment]
> And this leads to another concern -- memory footprint of the Table objects referenced by the cache.

These `table` objects would be mostly there already in `io.prestosql.plugin.hive.metastore.cache.CachingHiveMetastore#tableCache`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java (hunk 3 above)
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: Ok. Why?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 5]
```diff
@@ -1980,6 +1980,12 @@ private static void renameDirectory(HdfsContext context, HdfsEnvironment hdfsEnv
         return Optional.ofNullable(partition.getParameters().get(PRESTO_QUERY_ID_NAME));
     }
 
+    private Table getTableOrElseThrow(HiveIdentity identity, String databaseName, String tableName)
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: There would be name conflict then with other `getTable` method

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 4 above)
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java]: > What if ! partitionName.getKey().getTable().isPresent()?

It can't be empty as partitions cache always set `table` field.

> What if it is stale?

I think the assumption is that within transaction metadata doesn't really change. Otherwise exception should be thrown. Anyway, in current code metadata might be staled too (between `CachingHiveMetastore` and `ThriftMetastoreClient`)

**Comment by electrum** [comment]
Is it possible to add a test for the Avro schema change? Do we have tests for the other variant?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java [hunk 6]
```diff
@@ -94,11 +96,11 @@ public void createEmptyPartition(ConnectorSession session, String schema, String
         }
     }
 
-    private void doCreateEmptyPartition(ConnectorSession session, String schema, String table, List<Object> partitionColumnNames, List<Object> partitionValues)
+    private void doCreateEmptyPartition(ConnectorSession session, String schema, String tableName, List<Object> partitionColumnNames, List<Object> partitionValues)
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java]: Rename `schema` to be consistent

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 7]
```diff
@@ -610,30 +609,30 @@ private static boolean partitionValuesMatch(List<String> values, List<String> pa
         return true;
     }
 
-    public synchronized Optional<Partition> getPartition(HiveIdentity identity, String databaseName, String tableName, List<String> partitionValues)
+    public synchronized Optional<Partition> getPartition(HiveIdentity identity, Table table, List<String> partitionValues)
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: I think we can delete this method. If so, add this as a pre-commit

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 8]
```diff
@@ -868,8 +868,7 @@ public synchronized void revokeRoles(Set<String> roles, Set<HivePrincipal> grant
 
     private String getTableOwner(HiveIdentity identity, String databaseName, String tableName)
     {
-        Table table = delegate.getTable(identity, databaseName, tableName)
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: Can you do this refactoring change in a separate commit? It's hard to tell which ones are cleanup and which ones are new calls.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java [hunk 9]
```diff
@@ -178,15 +178,15 @@ private static AWSCredentialsProvider getAwsCredentialsProvider(GlueHiveMetastor
     {
         if (config.getAwsAccessKey().isPresent() && config.getAwsSecretKey().isPresent()) {
             return new AWSStaticCredentialsProvider(
-                new BasicAWSCredentials(config.getAwsAccessKey().get(), config.getAwsSecretKey().get()));
+                    new BasicAWSCredentials(config.getAwsAccessKey().get(), config.getAwsSecretKey().get()));
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java]: Please move formatting fix to a different commit

**Comment by electrum** [comment]
The second commit title is a bit long. Let's change to

> Remove redundant getTable call when fetching partitions

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java [hunk 10]
```diff
@@ -266,7 +266,15 @@ private void alterTable(HiveIdentity identity, String databaseName, String table
     @Override
     public Optional<Partition> getPartition(HiveIdentity identity, String databaseName, String tableName, List<String> partitionValues)
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java]: I think we should get rid of `getPartition` method, but we can do so afterwards

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java [hunk 11]
```diff
@@ -67,6 +78,12 @@ public HiveTableName getHiveTableName()
         return hiveTableName;
     }
 
+    @JsonProperty
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: Why does it need to be serialized?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java (hunk 3 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: shorter tostring

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 4 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java]: > It can't be empty as partitions cache always set `table` field.

then why do we need Optional?


> I think the assumption is that within transaction metadata doesn't really change.

note: this is not only about transactions, when caching enabled

Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java]: > then why do we need Optional?

`HivePartitionName` is used by different caches (e.g. stats) which provide different arguments or params. I expect this to simplify once we remove more metastore calls (e.g. for stats).

> note: this is not only about transactions, when caching enabled

In current code there might still be inconsistency (e.g. `getPartition` detects Avro format with table schema while `getTable` doesn't see that). Currently calls were leaking through session cache. This need to be fixed. If we find issues with long-term cache, we should consider changing strategy there.

**Comment by sopel39** [comment]
> Do we have tests for the getPartitionsByNames variant?

There is `TestAvroSchemaUrl#testPartitionedTableWithLongColumnType`.

> Is it possible to add a test for the Avro schema change?

I think the only sages of `getPartition` were during adding/updating of partition. I'm not sure I can repro that in Presto

**Comment by sopel39** [comment]
AC @electrum moved code to `HiveSplitManager` and `HiveWriteManager`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveUtil.java [hunk 12]
```diff
@@ -1018,4 +1020,14 @@ private static int getPositiveIntegerValue(Properties schema, String key, String
     {
         return toHiveTypes(schema.getProperty(IOConstants.COLUMNS_TYPES, ""));
     }
+
+    public static List<Column> getPartitionColumns(Table table, Partition partition)
+    {
+        boolean isAvroTableWithSchemaSet = isAvroTableWithSchemaSet(table);
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveUtil.java]: We can shorten this to
```java
return isAvroTableWithSchemaSet(table) ? table.getDataColumns() : partition.getColumns();
